pmlc/rt/runtime_registry.cc

```diff
@@ -37,54 +35,86 @@ class MemoizedFactory final {
   std::shared_ptr<Runtime> runtime;
 };
 
-const std::unordered_map<std::string, std::function<Runtime *()>>
-buildRuntimeMap() {
-  std::unordered_map<std::string, std::function<Runtime *()>> result;
+using Accessor = std::function<Runtime *()>;
 
-  auto &ll = LoaderList::instance();
-  std::lock_guard<std::mutex> lock{ll.mutex};
+class RuntimeRegistry {
+public:
+  static RuntimeRegistry *instance() {
+    static RuntimeRegistry reg;
+    return &reg;
+  }
 
-  for (auto &[loaderId, loader] : ll.loaders) {
-    std::unordered_map<std::string, Factory> factories;
-    try {
-      factories = loader();
-    } catch (const std::exception &e) {
-      IVLOG(1, "Loader " << loaderId << " initialization failed: " << e.what());
-      continue;
+  void registerFactory(llvm::StringRef id, Factory factory) {
+    std::lock_guard<std::mutex> lock{mutex};
+    auto memoizer = std::make_shared<Memoizer>(std::move(factory));
+    auto inserted =
+        runtimes
+            .emplace(id, Accessor{[memoizer = std::move(memoizer)]() {
+                       return memoizer->get();
+                     }})
+            .second;
+    if (!inserted) {
+      throw DuplicateRuntimeError{id};
     }
-    for (auto &[id, factory] : factories) {
-      auto memo = std::make_shared<MemoizedFactory>(factory);
-      auto [it, inserted] =
-          result.emplace(id, [memo = std::move(memo)]() -> Runtime * {
-            return memo->getRuntime();
-          });
-      if (!inserted) {
-        throw std::runtime_error{
-            llvm::formatv("Multiple runtime implementations found for {}", id)};
+  }
+
+  void bindRuntime(llvm::StringRef id, std::shared_ptr<Runtime> runtime) {
+    std::lock_guard<std::mutex> lock{mutex};
+    auto inserted =
+        runtimes
+            .emplace(id,
+                     [runtime = std::move(runtime)]() { return runtime.get(); })
+            .second;
+    if (!inserted) {
+      throw DuplicateRuntimeError{id};
+    }
+  }
+
+  Runtime *getRuntime(llvm::StringRef id) {
+    std::lock_guard<std::mutex> lock{mutex};
```

Comment:
I'm happy with marking these functions as not-thread-safe so we don't have to worry about locks.

Reply:
Done.